lib/libzfs/libzfs_sendrecv.c

```diff
@@ -3656,7 +3777,57 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 		nvlist_free(prop_errors);
 
 	if (newprops)
-		nvlist_free(props);
+		nvlist_free(rcvprops);
+
+	if (origprops != NULL)
+		nvlist_free(origprops);
+
+	return (err);
+}
+
+/*
+ * Check properties we were asked to override (both -o|-x)
+ */
+static int
+zfs_receive_checkprops(libzfs_handle_t *hdl, nvlist_t *props,
+    const char *errbuf)
+{
+	nvpair_t *nvp;
+	int err = 0;
+	const char *name;
+
+	nvp = NULL;
+	while ((nvp = nvlist_next_nvpair(props, nvp)) != NULL && err == 0) {
+		zfs_prop_t prop;
+		name = nvpair_name(nvp);
+		prop = zfs_name_to_prop(name);
+
+		if (prop == ZPROP_INVAL) {
+			if (zfs_prop_user(name) != B_TRUE) {
+				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+				    "invalid property '%s'"), name);
+				err = 1;
```

Comment:
seems like you can simplify this function by doing return (1) here, or preferably a boolean, and avoiding err variable altogether.